### PR TITLE
Added ability to look past incremental deploys

### LINKED ISSUE #5493: get_resource_events may leave events unhandled in case of incrmental deploy
This issue is purely theoretical and never observed. 

1. The idea of get_resource_events is `Get all events that should be processed by this specific resource, for the current deployment`
2. It does this by presenting all events since the last successful deploy.
3. However, an increment is ALSO a successful deploy, but it doesn't process events, so we might get a hole in the event processing coverage,..

I think this answer the question here: 

https://github.com/inmanta/inmanta-core/blob/5e7e7c887916f16497035daefbff218c351532e0/src/inmanta/server/services/resourceservice.py#L345

with yes

# Tasks
- [ ] implement event propagation part of the design: inmanta/designs#121

### PR DESCRIPTION
**Pull request by wouterdb**
# Description

Early review request to agree on the method used. 

closes  #5493

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )
- [ ] If this PR fixes a race condition in the test suite, also push the fix to the relevant stable branche(s) (see [test-fixes](https://internal.inmanta.com/development/core/tasks/build-master.html#test-fixes) for more info)

### CODE REVIEW

File: src/inmanta/data/__init__.py [hunk 1]
```diff
@@ -4256,66 +4256,43 @@ async def get_resource_events(
...
-            if result["begin_status"] != const.ResourceState.deploying:
-                raise BadRequest(
-                    "Fetching resource events only makes sense when the resource is currently deploying. Current deploy state"
-                    f" for resource {resource_version_id_str} is {result['begin_status']}."
-                )
-            current_deploy_start = result["begin_started"]
-            last_deploy_start = result["started"]
-
-            # Step3: Get the resource
+            print("STage 1")
```
Comment (wouterdb) [inline src/inmanta/data/__init__.py:4268]: todo

File: src/inmanta/data/__init__.py [hunk 2]
```diff
@@ -4256,66 +4256,43 @@ async def get_resource_events(
...
 
+            if resource.status != const.ResourceState.deploying:
+                raise BadRequest(
+                    "Fetching resource events only makes sense when the resource is currently deploying. Current deploy state"
+                    f" for resource {resource_version_id_str} is {resource.status}."
+                )
+
+            # steps 2:
+            # find the interval between the current deploy (now) and the previous successful deploy
+            print("Stage 2:", resource.last_success)
```
Comment (wouterdb) [inline src/inmanta/data/__init__.py:4285]: todo

File: src/inmanta/server/services/resourceservice.py [hunk 3]
```diff
@@ -582,6 +588,7 @@ async def resource_action_update(
         change: const.Change,
         send_events: bool,
         keep_increment_cache: bool = False,
+        is_increment_notification: bool = False,
```
Comment (wouterdb) [inline src/inmanta/server/services/resourceservice.py:591]: We may want to get rid of this method in iso7

**Review by arnaudsjs — CHANGES_REQUESTED** [review]
The method used looks good to me.

File: src/inmanta/data/__init__.py [hunk 4]
```diff
@@ -4967,6 +4946,24 @@ async def get_resources_in_resource_sets(
...
+        version: int,
+    ) -> None:
+        last_released = await ConfigurationModel.get_latest_version(environment)
+        if not last_released:
+            return
+        previous_version = last_released.version
+        query = (
+            f"UPDATE {cls.table_name()} as new_resource "
+            f"SET last_success = (SELECT last_success from {cls.table_name()} as old_resource where old_resource.model=$3 and old_resource.environment=$2 and old_resource.resource_id=new_resource.resource_id)"
+            f"where new_resource.model=$1 and new_resource.environment=$2 and new_resource.last_success is null"
```
[the remark this answers is not in the document]
Comment (wouterdb) [reply src/inmanta/data/__init__.py:4962]: I tried to improve it, but I don't know the convention

File: src/inmanta/data/__init__.py [hunk 5]
```diff
@@ -4157,6 +4157,7 @@ async def query_resource_actions(
         first_timestamp: Optional[datetime.datetime] = None,
         last_timestamp: Optional[datetime.datetime] = None,
         action: Optional[const.ResourceAction] = None,
+        resource_id: Optional[ResourceIdStr] = None,
```
Comment (wouterdb) [inline src/inmanta/data/__init__.py:4160]: I don't like this method breaking up the id's so I updated it to be able to handle complete id's

File: tests/db/migration_tests/test_v202306060_to_v202308010.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+            # verify time on last success
+            actions = await ResourceAction.query_resource_actions(
+                environment=env.id, resource_id=resource.resource_id, action=const.ResourceAction.deploy
+            )
+            last_deploy = actions[0]
+            last_deploy.version == 2
+            assert resource.last_success == last_deploy.finished
+            expected += 1
+        if resource.resource_id == "std::File[localhost,path=/tmp/test]":
+            # always fails
```
Comment (arnaudsjs) [inline tests/db/migration_tests/test_v202306060_to_v202308010.py:54]: Can you elaborate on why this one always fails?

File: src/inmanta/db/versions/v202308010.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    """
+    await connection.execute(schema)
+
+    # For each environment
+    # For the latest released version
+    # Set the last success
+    update_query = """
+    UPDATE resource as r
+    SET last_success = (
+        SELECT max(started)
```
Comment (arnaudsjs) [inline src/inmanta/db/versions/v202308010.py:31]: Why do we use `started` here instead of finished?

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: src/inmanta/db/versions/v202308010.py (hunk 7 above)
Comment (wouterdb) [reply src/inmanta/db/versions/v202308010.py:31]:
```suggestion
        SELECT max(finished)
```

**Review by arnaudsjs — APPROVED** [review]

File: changelogs/unreleased/5493-get_events_accross_incremental_deploys.yml [hunk 8]
```diff
@@ -0,0 +1,6 @@
+description: Ensure get_resource_events works across incremental compiles
```
Comment (sanderr) [inline changelogs/unreleased/5493-get_events_accross_incremental_deploys.yml:1]:
```suggestion
description: Ensure get_resource_events and resource_did_dependency_change work across incremental compiles
```

File: src/inmanta/data/__init__.py [hunk 9]
```diff
@@ -4256,66 +4261,41 @@ async def get_resource_events(
...
 
+            if resource.status != const.ResourceState.deploying:
+                raise BadRequest(
+                    "Fetching resource events only makes sense when the resource is currently deploying. Current deploy state"
+                    f" for resource {resource_version_id_str} is {resource.status}."
+                )
+
+            # steps 2:
+            # find the interval between the current deploy (now) and the previous successful deploy
+            last_deploy_start = resource.last_success
```
Comment (sanderr) [inline src/inmanta/data/__init__.py:4289]: 1. Might `last_successful_execute` be a better name? To make clear that it's not just about success "due to known good state" but explicit success due to handler execution? Completely optional suggestion.
2. As discussed, if we make this the metric to determine the window within which we look at events, it should be a part of these methods' contract/docstring and the handler's reload documentation. Perhaps with an explicit clarification that this means that a handler that uses these methods to monitor or respond to dependencies' events should mark its resource as deployed iff it successfully and completely processed all current events.

File: src/inmanta/server/services/resourceservice.py [hunk 10]
```diff
@@ -539,10 +540,15 @@ async def resource_deploy_done(
                 if not keep_increment_cache:
                     self.clear_env_cache(env)
 
+                extra_fields = {}
+                if status == ResourceState.deployed:
+                    extra_fields["last_success"] = finished
```
Comment (sanderr) [inline src/inmanta/server/services/resourceservice.py:545]: If we use `finished` here, don't we risk missing events between when `get_resource_events` was called and when `resource_deploy_done` is called? I'd think we either need to use `started` (meaning we also need to revert the change you made to the filter higher up) or allow the handler to explicitly register the time at which it looked at events (which would also require an update of the filter).

File: src/inmanta/db/versions/v202308010.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    """
+    await connection.execute(schema)
+
+    # For each environment
+    # For the latest released version
+    # Set the last success
+    update_query = """
+    UPDATE resource as r
+    SET last_success = (
+        SELECT max(finished)
```
Comment (sanderr) [inline src/inmanta/db/versions/v202308010.py:31]: Related to other comment about using `finished` as marker. Though in this case it may be our best option regardless.

File: src/inmanta/server/services/resourceservice.py (hunk 3 above)
Comment (sanderr) [inline src/inmanta/server/services/resourceservice.py:591]: Could you add this to the method docstring?

**Review by sanderr — CHANGES_REQUESTED** [review]

File: src/inmanta/data/__init__.py (hunk 9 above)
Comment (wouterdb) [reply src/inmanta/data/__init__.py:4289]: for point 2, I added some docs to the relevant api endpoint, but it was already quite good

for reload, I'll wait for https://github.com/inmanta/inmanta-core/pull/6246

File: src/inmanta/db/versions/v202308010.py (hunk 11 above)
Comment (wouterdb) [reply src/inmanta/db/versions/v202308010.py:31]: modified

File: tests/db/migration_tests/test_v202306060_to_v202308010.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+    for resource in resources:
+        if resource.resource_id == "std::AgentConfig[internal,agentname=localhost]":
+            # always success
+            assert resource.last_success
+            # verify time on last success
+            actions = await ResourceAction.query_resource_actions(
+                environment=env.id, resource_id=resource.resource_id, action=const.ResourceAction.deploy
+            )
+            last_deploy = actions[1]
+            assert last_deploy.version == 2
```
Comment (wouterdb) [inline tests/db/migration_tests/test_v202306060_to_v202308010.py:50]: I had a missing assert here, which allowed me to find a bug

File: src/inmanta/server/services/resourceservice.py [hunk 13]
```diff
@@ -592,6 +592,12 @@ async def resource_action_update(
         *,
         connection: Optional[Connection] = None,
     ) -> Apireturn:
+        """
+        :param is_increment_notification: is this the increment calucation setting the deployed status,
+            instead of an actual deploy? Used to keep track of the last_success field on the resources,
+            which should not be updated for increments.
```
Comment (sanderr) [inline src/inmanta/server/services/resourceservice.py:598]: Shouldn't this be documented on `methods.resource_action_update` instead?

File: src/inmanta/data/__init__.py (hunk 9 above)
Comment (sanderr) [reply src/inmanta/data/__init__.py:4289]: > I added some docs to the relevant api endpoint

Did you push this?

**Review by sanderr — CHANGES_REQUESTED** [review]

File: src/inmanta/data/__init__.py (hunk 9 above)
Comment (wouterdb) [reply src/inmanta/data/__init__.py:4289]: I did not

File: src/inmanta/server/services/resourceservice.py (hunk 13 above)
Comment (wouterdb) [reply src/inmanta/server/services/resourceservice.py:598]: no, this argument doens't exist on the rest api, only internally.

**Review by sanderr — CHANGES_REQUESTED** [review]
Still [one](https://github.com/inmanta/inmanta-core/pull/6344#discussion_r1284212187) unresolved comment.

**Review by sanderr — APPROVED** [review]

**Comment by sanderr** [comment]
> Still https://github.com/inmanta/inmanta-core/pull/6344#discussion_r1284212187 unresolved comment.

Race condition
